Modules/_collectionsmodule.c

```diff
@@ -1988,6 +2116,28 @@ dequereviter_next(dequeiterobject *it)
     return Py_NewRef(item);
 }
 
+static PyObject *
+dequereviter_next(dequeiterobject *it)
+{
+    dequeobject *deque = NULL;
+    Py_ssize_t counter;
+
+    Py_BEGIN_CRITICAL_SECTION(it);
+    counter = it->counter;
+    deque = it->deque;
+    Py_END_CRITICAL_SECTION();
+
+    if (counter == 0)
+        return NULL;
```

Comment:
I think we should remove this critical section and push the `counter == 0` check inside `dequereviter_next_lock_held`. That function needs to handle the `counter == 0` case anyways. It's not safe to check it inside a previous critical section and assume it's still true in a later critical section.